### PR TITLE
Add LocalDateTime type definition into Core vocabulary

### PR DESCRIPTION
**Pull request by xuzhg**
### Issues

* This pull request add the "LocalDateTime" type definition into Core vocabulary.

See PR at: https://github.com/oasis-tcs/odata-vocabularies/pull/23

### Description

* OData TC has a update for Core vocabulary by adding "LocalDateTime" type definition. see: oasis-tcs/odata-vocabularies#23

* This PR is to update the core type definition in the Edm library.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Edm/Vocabularies/CoreVocabularies.xml [hunk 1]
```diff
@@ -310,4 +310,8 @@ Any simple identifier | Any type listed in `Validation.OpenPropertyTypeConstrain
     <Annotation Term="Core.LongDescription" String="The annotation value will usually be an expression, e.g. using properties of the binding parameter type for instance-dependent availability, or using properties of a singleton for global availability. The static value `null` means that availability cannot be determined upfront and is instead expressed as an operation advertisement." />
   </Term>
 
+  <TypeDefinition Name="LocalDateTime" UnderlyingType="Edm.String">
+    <Annotation Term="Core.Description" String="A string representing a Local Date-Time value with no offset." />
```
Comment (KanishManuja-MS) [inline src/Microsoft.OData.Edm/Vocabularies/CoreVocabularies.xml]: Does http://odata.github.io/WebApi/#04-01-datetime-support need to be updated?

Comment (xuzhg) [reply src/Microsoft.OData.Edm/Vocabularies/CoreVocabularies.xml]: That's still true. OData V4 spec doesn't include "DateTime" as built-in primitive type.

It's the next step to use the "LocalDateTime" type definition in Web API.

Comment (mikepizzo) [reply src/Microsoft.OData.Edm/Vocabularies/CoreVocabularies.xml]: exactly.  LocalDateTime is not an actual type; its type is string. The typedef puts constraints on the format of the string, and defines semantics around what the string represents.

---
In reply to: [229088414](https://github.com/OData/odata.net/pull/1308#discussion_r229088414) [](ancestors = 229088414)

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs [hunk 2]
```diff
@@ -961,6 +961,41 @@ public void ShouldWriteAnnotationForEnumMember()
...
+            Assert.NotNull(qualifiedTypeName);
+
+            EdmComplexType type = new EdmComplexType("NS", "Complex");
+            type.AddStructuralProperty("ModifiedDate", new EdmTypeDefinitionReference(localDateTime, true));
+            type.AddStructuralProperty("QualifiedName", new EdmTypeDefinitionReference(qualifiedTypeName, true));
+
+            model.AddElement(type);
+            IEnumerable<EdmError> errors;
+            Assert.False(model.Validate(out errors));
+            Assert.Equal(2, errors.Count());
```
Comment (mikepizzo) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs]: >2 [](start = 25, length = 1)

Why does validation fail for LocalDateTime, if it's added to the Core vocabulary?

Comment (xuzhg) [reply test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs]: Failed for the Edm validation rule, the error message is:
`
{InvalidPropertyType : A property cannot be of type 'TypeDefinition'. The property type must be a complex, a primitive, an enum or an untyped type, or a collection of complex, primitive, or enum types.}
`
I thought that's a wrong validatation rule. I also sent it through Email to you. Please take a look.

[the remark this answers is not in the document]
Comment (xuzhg) [reply test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs]: That's related to the validation rule. We should modify it in other issue/PR.

See detail comments in below @mikepizzo .

[the remark this answers is not in the document]
Comment (xuzhg) [reply test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs]: @biaol-odata 

Copy the error message is:
{InvalidPropertyType : A property cannot be of type 'TypeDefinition'. The property type must be a complex, a primitive, an enum or an untyped type, or a collection of complex, primitive, or enum types.}

**Comment by xuzhg** [comment]
@mikepizzo @biaol-odata I created the issue for the property with type-definition type validation. See https://github.com/OData/odata.net/issues/1312.

**Comment by xuzhg** [comment]
@mikepizzo @biaol-odata I fixed the issue #1312 in this PR's second commit. Please take a look and let me any question.

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs [hunk 3]
```diff
@@ -961,6 +961,40 @@ public void ShouldWriteAnnotationForEnumMember()
...
+            var qualifiedTypeName = model.FindType("Org.OData.Core.V1.QualifiedTypeName") as IEdmTypeDefinition;
+            Assert.NotNull(qualifiedTypeName);
+
+            EdmComplexType type = new EdmComplexType("NS", "Complex");
+            type.AddStructuralProperty("ModifiedDate", new EdmTypeDefinitionReference(localDateTime, true));
+            type.AddStructuralProperty("QualifiedName", new EdmTypeDefinitionReference(qualifiedTypeName, true));
+
+            model.AddElement(type);
+            IEnumerable<EdmError> errors;
+            Assert.True(model.Validate(out errors));
```
Comment (mikepizzo) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs]: should you add an Assert.Equal(0,errors.Count())?

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikepizzo) [reply test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs]: Resolved by fixing rule.

---
In reply to: [229124840](https://github.com/OData/odata.net/pull/1308#discussion_r229124840) [](ancestors = 229124840)

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs (hunk 3 above)
Comment (xuzhg) [reply test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs]: Thanks. Normally, it's unnecessary.
 `Validate` method just likes "Try and Do something". Return `true` means there's no errors.

In the inner, this `validate` method is returning using the code like

 `return errors.FirstOrDefault() == null;`

**Pull request merged**